### PR TITLE
Add download entire period bundle feature

### PR DESCRIPTION
**Pull request by maxneuvians**
This is a feature that will allow a person to download all the key for the period defined by `numberOfDaysToServe`, currently 14, in one package.

The access the feature,` enableEntirePeriodBundle ` needs to be set to `true`. 

Instead of the regular day number since unix time, ex: `18440` we use `00000` because that is not a valid date but does not require the creation of a new API endpoint or changing the authorization logic behind retrieving bundles. The mobile client can use that as a default before replacing it with the current day.

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -25,6 +25,7 @@ type Constants struct {
 	HmacKeyLength                  			int
 	CORSAccessControlAllowOrigin   			string
 	DisableCurrentDateCheckFeatureFlag	bool
+	EnableEntirePeriodBundle						bool
```
Comment (sboisvert) [inline pkg/config/config.go:28]: Super minor, and maybe this is just the github editor, but it doesn't line up with the others. Do we have a linter / formatter?

File: pkg/server/retrieve.go [hunk 2]
```diff
@@ -68,24 +68,42 @@ func (s *retrieveServlet) retrieve(w http.ResponseWriter, r *http.Request) resul
...
+
+	if config.AppConstants.EnableEntirePeriodBundle == true && vars["day"] == "00000" {
+
+		endDate := timemath.CurrentDateNumber() - 1
+		startDate := endDate - numberOfDaysToServe
+
+		dateNumber = endDate
+
+		startTimestamp = time.Unix(int64(startDate*86400), 0)
+		endTimestamp = time.Unix(int64((endDate+1)*86400), 0)
```
Comment (sboisvert) [inline pkg/server/retrieve.go:83]: Maybe I'm reading this wrong, but it seems like we're doing "yesterday" on L77, then taking numberOfDaysToServe in the past, and then on L83 we're adding back the day we removed so we're back to "today" 
This seems to imply that we're also serving today as part of the bundle, is there ever a use case where we'd want DisableCurrentDateCheckFeatureFlag to be false and EnableEntirePeriodBundle to be true? It seems like we're assuming that EnableEntirePeriodBundle means DisableCurrentDateCheckFeatureFlag is also true.
It's also entirely possible that my timeMath is just bad.

File: pkg/server/retrieve.go [hunk 3]
```diff
@@ -68,24 +68,42 @@ func (s *retrieveServlet) retrieve(w http.ResponseWriter, r *http.Request) resul
...
+		endTimestamp = time.Unix(int64((dateNumber+1)*86400), 0)
+
+	}
 
 	currentRSIN := pb.CurrentRollingStartIntervalNumber()
 	currentDateNumber := timemath.CurrentDateNumber()
 
 	if config.AppConstants.DisableCurrentDateCheckFeatureFlag == false && dateNumber == currentDateNumber {
-		return s.fail(log(ctx, err), w, "request for current date", "cannot serve data for current period for privacy reasons", http.StatusNotFound)
+		return s.fail(log(ctx, nil), w, "request for current date", "cannot serve data for current period for privacy reasons", http.StatusNotFound)
```
Comment (sboisvert) [inline pkg/server/retrieve.go:102]: Got it, we're checking it here, so we do need both to be true for it to work.

File: test/retrieve_test.rb [hunk 4]
```diff
@@ -87,6 +87,46 @@ def test_retrieve_stuff
     assert_keys(export, keys, region: 'CA', date_number: dn)
   end
 
+  def test_all_keys
+    active_at = time_in_date('10:00', today_utc.prev_day(8))
+    two_days_ago = yesterday_utc.prev_day(1)
+
+    # Our retrieve endpoint returns keys SUBMITTED within the given period.
```
Comment (sboisvert) [inline test/retrieve_test.rb:94]: Should we check the bounds? Like 14 days ago and today as well?

File: pkg/server/retrieve.go (hunk 2 above)
Comment (maxneuvians) [reply pkg/server/retrieve.go:83]: endTimestamp is just the outer bound of the period, so you go up to the end days midnight. We need to ensure that yesterdays day is used in `startDate` and `dateNumber` and that the correct endTimestamp is then calculated.

File: pkg/config/config.go (hunk 1 above)
Comment (maxneuvians) [reply pkg/config/config.go:28]: Fixed

File: pkg/server/retrieve.go (hunk 3 above)
Comment (maxneuvians) [reply pkg/server/retrieve.go:102]: 👍

File: test/retrieve_test.rb (hunk 4 above)
Comment (maxneuvians) [reply test/retrieve_test.rb:96]: Added!

**Comment by henrytao-me** [comment]
Link https://github.com/cds-snc/covid-shield-mobile/issues/642

**Pull request merged**